src/sinks/pulsar/sink.rs

```diff
@@ -0,0 +1,93 @@
+use async_trait::async_trait;
+
+use futures::{future, stream::BoxStream, StreamExt};
+use pulsar::{Error as PulsarError, Pulsar, TokioExecutor};
+use snafu::{ResultExt, Snafu};
+use tower::limit::ConcurrencyLimit;
+use vector_core::config::log_schema;
+use vector_core::event::LogEvent;
+use vector_core::sink::StreamSink;
+
+use crate::sinks::pulsar::config::{PulsarSinkConfig, QUEUED_MIN_MESSAGES};
+use crate::sinks::pulsar::request_builder::PulsarRequestBuilder;
+use crate::sinks::pulsar::service::PulsarService;
+use crate::sinks::util::SinkBuilderExt;
+use crate::template::{Template, TemplateParseError};
+use crate::{
+    codecs::{Encoder, Transformer},
+    event::Event,
+};
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub(crate)))]
+pub(crate) enum BuildError {
+    #[snafu(display("creating pulsar producer failed: {}", source))]
+    CreatePulsarSink { source: PulsarError },
+    #[snafu(display("invalid topic template: {}", source))]
+    TopicTemplate { source: TemplateParseError },
+}
+
+pub(crate) struct PulsarSink {
+    transformer: Transformer,
+    encoder: Encoder<()>,
+    service: PulsarService<TokioExecutor>,
+    topic: Template,
+    key_field: Option<String>,
+    properties_key: Option<String>,
+}
+
+pub(crate) async fn healthcheck(config: PulsarSinkConfig) -> crate::Result<()> {
+    trace!("Healthcheck started.");
+    let client = config.create_pulsar_client().await?;
+    let topic = Template::try_from(config.topic)
+        .context(TopicTemplateSnafu)?
+        .render_string(&LogEvent::from_str_legacy(""))?;
+    client.lookup_topic(topic).await?;
+    trace!("Healthcheck completed.");
+    Ok(())
+}
+
+impl PulsarSink {
+    pub(crate) fn new(
+        client: Pulsar<TokioExecutor>,
+        config: PulsarSinkConfig,
+    ) -> crate::Result<Self> {
+        let producer_opts = config.build_producer_options();
+        let transformer = config.encoding.transformer();
+        let serializer = config.encoding.build()?;
+        let encoder = Encoder::<()>::new(serializer);
+        let service = PulsarService::new(client, producer_opts, None);
+
```

Comment:
I believe you could do the same as this code is, by utilizing the `ServiceBuilder` to apply the `TowerRequestSettings`

https://github.com/vectordotdev/vector/blob/8fd45eb09c177b1ced9102ba693b7f7699ecd243/src/sinks/aws_kinesis_streams/config.rs#L171-L180